File: src/sc_neurocore/accel/julia/sleep/circadian_optimizer.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for sleep/circadian_optimizer

module CircadianOptimizerAccel

using Statistics, LinearAlgebra

mutable struct CircadianOptimizerState
    chronotype::Float64
    bedtime_hour::Float64
    wake_hour::Float64
    default_protocol::Float64
    melatonin_peak_hour::Float64
    core_body_temp_nadir_hour::Float64
    _profile::Float64
end

function CircadianOptimizerState()
    CircadianOptimizerState(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
end

function get_profile(s::CircadianOptimizerState)
    return s._profile
end

function get_sleep_window(s::CircadianOptimizerState)
    return (s._profile.bedtime_hour, s._profile.wake_hour)
end

function get_recommended_protocol(s::CircadianOptimizerState)
    return s._profile.default_protocol
end

function is_in_sleep_window(s::CircadianOptimizerState, hour)
    bed = s._profile.bedtime_hour
    wake = s._profile.wake_hour
    if bed <= wake
        return bed <= hour < wake
    else
        # wraps past midnight
        return hour >= bed || hour < wake
end

function melatonin_level(s::CircadianOptimizerState, hour)
    peak = s._profile.melatonin_peak_hour
    # phase so that cos(0) = 1 at the peak hour
    phase = 2.0 * math.pi * (hour - peak) / 24.0
    level = 0.5 * (1.0 + math.cos(phase))
    return float(clamp(level, 0.0, 1.0))
end

function to_dict(s::CircadianOptimizerState)
    p = s._profile
    return {
        "chronotype": s.chronotype.value,
        "bedtime_hour": p.bedtime_hour,
        "wake_hour": p.wake_hour,
        "default_protocol": p.default_protocol,
        "melatonin_peak_hour": p.melatonin_peak_hour,
        "core_body_temp_nadir_hour": p.core_body_temp_nadir_hour,
    }
end

end # module CircadianOptimizerAccel

File: src/sc_neurocore/accel/rust/safety/circadian_optimizer.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for circadian_optimizer

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct CircadianOptimizer {
    pub chronotype: f64,
    pub bedtime_hour: f64,
    pub wake_hour: f64,
    pub default_protocol: f64,
    pub melatonin_peak_hour: f64,
    pub core_body_temp_nadir_hour: f64,
    pub _profile: f64,
}

impl CircadianOptimizer {
    pub fn new() -> Self {
        Self {
            chronotype: 0.0_f64,
            bedtime_hour: 0.0_f64,
            wake_hour: 0.0_f64,
            default_protocol: 0.0_f64,
            melatonin_peak_hour: 0.0_f64,
            core_body_temp_nadir_hour: 0.0_f64,
            _profile: 0.0_f64,
        }
    }

    pub fn get_profile(&self, ) -> f64 {
        // return self._profile
        0.0
    }

    pub fn get_sleep_window(&self, ) -> f64 {
        // return (self._profile.bedtime_hour, self._profile.wake_hour)
        0.0
    }

    pub fn get_recommended_protocol(&self, ) -> f64 {
        // return self._profile.default_protocol
        0.0
    }

    pub fn is_in_sleep_window(&self, hour: f64) -> f64 {
        // bed = self._profile.bedtime_hour
        // wake = self._profile.wake_hour
        // if bed <= wake:
        // return bed <= hour < wake
        // else:
        // # wraps past midnight
        // return hour >= bed || hour < wake
        0.0
    }

    pub fn melatonin_level(&self, hour: f64) -> f64 {
        // peak = self._profile.melatonin_peak_hour
        // # phase so that cos(0) = 1 at the peak hour
        // phase = 2.0 * math.pi * (hour - peak) / 24.0
        // level = 0.5 * (1.0 + math.cos(phase))
        // return float((level_f64).clamp(0.0, 1.0))
        0.0
    }

    pub fn to_dict(&self, ) -> f64 {
        // p = self._profile
        // return {
        // "chronotype": self.chronotype.value,
        // "bedtime_hour": p.bedtime_hour,
        // "wake_hour": p.wake_hour,
        // "default_protocol": p.default_protocol,
        // "melatonin_peak_hour": p.melatonin_peak_hour,
        // "core_body_temp_nadir_hour": p.core_body_temp_nadir_hour,
        // }
        0.0
    }

}

pub fn validate_circadian_optimizer(state: &CircadianOptimizer) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circadian_optimizer_new() {
        let state = CircadianOptimizer::new();
        assert!(validate_circadian_optimizer(&state));
    }

}
